### DISCUSSION TITLE
Message-level metadata

### DISCUSSION THREAD

**Discussion by nkmnz** [chatwoot/chatwoot#8188] [Feature Requests]
### Is your feature or enhancement related to a problem? Please describe.

We use chatwoot for customer support on a platform with highly individualized content. In order to give equally targeted support via chat, we need to deliver the exact context of the user's experience to the support staff. Since there's always some delay between a customer's chat request and the answer, it's not enough to sync the current state of the webapp with the conversation, but we rather need to store a snapshot of the current state with each message. Right now, the only way we could figure out to this is to set this "snapshot" with the metadata of the conversation, forward each message via a webhook from chatwoot to our backend, and store each message next to the metadata. The support can then access the metadata of the conversation at the time of the message through our systems, which gives them the necessary context of each individual message.   

### Describe the solution you'd like

There should be a way to set metadata that is shown in the support interface within the conversation. It could look similar to an individual message from the user, but it shouldn't be shown to the user in the widget UI. The most important difference to conversation-level metadata would be that they are not overridden, but appended to (or included into) the ongoing conversation. 
We could call it a "Signal". Simple usecase for illustration: an ecommerce store could log the products which a customer is looking at during a consultation with a chat agent. Or we could log if a user has clicked on a link that support sent them. Or, like in our case, what kind of content they are currently seeing.

**Comment by pranavrajs** [comment]
@nkmnz Thanks for sharing your thoughts. I'm moving this to the feature discussions as I couldn't figure out an immediate action item from the description.

A couple of follow-up questions:

1/ Why is it required to have the metadata tagged with a message? Is the metadata captured when the message is sent? I don't understand how we can support it in the widget SDK.
2/ I see the use case to be more of an events-like item. For eg: If I connect an analytics tool and identify the user, I can see the actions they have taken on the website. To get that context in Chatwoot, you can the [Dashboard Apps](https://www.chatwoot.com/docs/product/others/dashboard-apps/) and embed it into the support agent dashboard. Does this solution work?

**Reply by nkmnz** [reply]
Hi @pranavrajs , thank you for taking the time to look into this! You are right: in one sentence, it boils down to showing external data, e.g. from product analytics, in the context of the conversation. From this point of view, we could just use the timestamp from a message and pull the data from our backend into a dashboard app view. I'd still prefer to show the events within the conversation (toggle on/off and filter by event type). This could be awesome way to display data from tools like posthog/mixpanel or stripe.

**Comment by alchemistake** [comment]
we need this on our company as well
